### PR TITLE
Expose response object to callback

### PR DESCRIPTION
**Pull request by djmcdonald**

### CODE REVIEW

**Comment by djmcdonald** [comment]
Fixes a few jshint errors and allows us to examine the response object (could get a number of potential success codes that we need to do different things to).

**Comment by robinjmurphy** [comment]
Sounds good. Are you able to update the README? I'll give this a test out locally tomorrow.

File: test/client.js [hunk 1]
```diff
@@ -432,6 +432,17 @@ describe('Rest Client', function () {
 
       client.get(url, done);
     });
+
+    it('should pass the response object through', function (done) {
```
Comment (robinjmurphy) [inline test/client.js]: Can you update these to be say `passes the response object through` without the `should`? Just to match the styles of the other tests.

**Comment by robinjmurphy** [comment]
We'll need some tests around caching. The response object will need to be returned when responses are delivered from the cache.

**Comment by djmcdonald** [comment]
Good point - I'll look at adding that tomorrow: I'll update the readme also.

**Comment by djmcdonald** [comment]
@robinjmurphy before I implement the caching side of the response I thought I'd run something past you: the response object generated by request is quite large and probably contains things that most users won't care about (it also contains non serialisable things like functions). I was thinking that for now I would cache:
 * the response code
 * response headers
 * maybe a `isCached` property

If you're OK with this I'll implement it and update the readme accordingly.

**Comment by robinjmurphy** [comment]
This is the hard part. I think the best thing would be to return the entire familiar response object rather than creating our own, but I appreciate that the response isn't serialisable.

Maybe the best we can do is return a new response object as the third argument to the callback with a subset of properties. The `statusCode` and `headers` sound like a good start. I don't think we need `isCached`. Whether the response came from cache or not should be transparent.

Maybe worth including `body` as a property in the response object too.

**Comment by djmcdonald** [comment]
@robinjmurphy I've opted to store the status code and headers for now. 

I'm wondering if it is worth checking the full response object is passed through if nothing exists in cache for it yet? However, given a resp object wasn't available just having the status code and headers is a start.

I've updated the readme as well - would you mind re-reviewing please?

**Comment by robinjmurphy** [comment]
The outstanding bit for me is that I think we should return the same 'subset' response object regardless of whether the response comes out of the cache or not. Otherwise you could end up writing code that works fine when the response is fresh but starts failing when it comes out of the cache. If, for example, you assume the presence of a property on the response that isn't in the cached version. I'd like to keep the return values the same in both the cached and un-cached state if possible.

Other than that, would be interested to see what @DaMouse404 and @nspragg reckon to the general idea.

**Comment by djmcdonald** [comment]
Agreed. I'll write tests/code for that.

**Comment by djmcdonald** [comment]
@robinjmurphy made that change and updated the readme to reflect - sorry it took so long.

**Comment by robinjmurphy** [comment]
Thanks. @DaMouse404 would be interested to get your take on this if you could spare some time?

**Comment by nspragg** [comment]
I think returning a generic response object is fine (keeping it simple to start). This is better than returning a response which is coupled to a specific implementation (like request). 

Agree with @robinjmurphy  - caching should be transparent i.e no need to isCached && response objects should be consistent whether cached or not.

File: lib/cachingClient.js [hunk 2]
```diff
@@ -82,18 +82,24 @@ CachingClient.prototype.get = function (url, opts, cb) {
         cachedObject.body = body;
       }
 
+      if (res) {
```
Comment (nspragg) [inline lib/cachingClient.js]: This if statement is not required. It's already tested in the surrounding if.

File: lib/client.js [hunk 3]
```diff
@@ -99,7 +99,16 @@ Client.prototype._request = function (method, url, opts, cb) {
   });
 
   request[method](url, opts, function (err, res, body) {
-    var statsName;
+    var statsName,
```
Comment (nspragg) [inline lib/client.js]: For consistency please use separate var declarations.

File: lib/client.js [hunk 4]
```diff
@@ -99,7 +99,16 @@ Client.prototype._request = function (method, url, opts, cb) {
...
+        cacheableRes = {};
+
+    if (res) {
```
Comment (nspragg) [inline lib/client.js]: If we move this after the current early return for the error then there's no need to check for a `res`.

File: lib/client.js [hunk 5]
```diff
@@ -99,7 +99,16 @@ Client.prototype._request = function (method, url, opts, cb) {
...
+      cacheableRes = {
```
Comment (nspragg) [inline lib/client.js]: Is there another name we could use for `cacheableRes`? Maybe call them `requestRes` and `flashheartRes`?

**Comment by djmcdonald** [comment]
@nspragg I've made the requested changes - I'm not sure `flashheartRes` is an appropriate variable name. I chose `cacheableRes` as I felt it somewhat describes why I am cutting the result down to a sub-set.

**Comment by djmcdonald** [comment]
hey @robinjmurphy was just wondering if you had any further comments for this PR?

**Comment by robinjmurphy** [comment]
Hi @djmcdonald. Apologies for the delay! We're going to get this merged and released today.

**Comment by djmcdonald** [comment]
Great, thanks @robinjmurphy !

**Comment by robinjmurphy** [comment]
@djmcdonald No worries. Ping me where you sit by email and I'll send you up a lovely Flashheart laptop sticker (they're actually a thing.)

**Pull request merged**